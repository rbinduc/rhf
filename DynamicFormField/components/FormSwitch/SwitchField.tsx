import React, {ReactNode} from 'react';
import Switch, {SwitchProps} from '@material-ui/core/Switch';
import styled from '@emotion/styled';
import {ITenantTheme} from '../../../../../../models';
import {ControllerRenderProps, FieldValues} from 'react-hook-form';

export interface ISwitchFieldProps extends SwitchProps {
  label?: ReactNode;
  ariaLabel?: string;
  field?: ControllerRenderProps<FieldValues, any>;
  isDirty?: boolean;
  isDisabled?: boolean;
  checked?: boolean;
}

const SwitchField = ({label, checked, field, isDisabled, size, ariaLabel, ...rest}: ISwitchFieldProps) => {
  return (
    <StyledSwitch
      {...rest}
      inputProps={{
        'aria-label': ariaLabel ?? (label && typeof label === 'string' ? label : ''),
        'aria-hidden': isDisabled,
        ...rest.inputProps,
      }}
      checked={checked}
      disabled={isDisabled}
      size={size}
      onChange={(event) => {
        if (field) {
          field?.onChange((event.target as HTMLInputElement).checked);
        } else {
          rest?.onChange && rest.onChange(event, (event.target as HTMLInputElement).checked);
        }
      }}
    />
  );
};

type StyledSwitchCSS = {
  rootWidth?: string;
  rootHeight?: string;
  rootPadding?: string;
  switchOnLabel?: string;
  switchOffLabel?: string;
  theme?: ITenantTheme;
};

const StyledSwitch = styled(Switch)<StyledSwitchCSS>`
  .MuiSwitch-root {
    width: ${(props) => props.rootWidth};
    height: ${(props) => props.rootHeight};
  }
  .MuiSwitch-switchBase {
    color: ${({
      theme: {
        designSystem: {
          organisms: {
            formSwitch: {switchBaseColor = ''},
          },
        },
      },
    }) => switchBaseColor ?? ''};
  }
  .MuiSwitch-colorSecondary.Mui-checked + .MuiSwitch-track {
    background-color: ${({
      theme: {
        designSystem: {
          organisms: {
            formSwitch: {checkedBackgroundColor = ''},
          },
        },
      },
    }) => checkedBackgroundColor ?? ''};
  }
  .MuiSwitch-track {
    background-color: ${({
      theme: {
        designSystem: {
          organisms: {
            formSwitch: {unCheckedBackgroundColor = ''},
          },
        },
      },
    }) => unCheckedBackgroundColor ?? ''};;
    &:before {
      content: '${(props) => props.switchOnLabel}'
    }
    &:after {
      content: '${(props) => props.switchOffLabel}'
    }
  }
  .MuiSwitch-thumb {
    border: ${({
      theme: {
        designSystem: {
          organisms: {
            formSwitch: {unCheckedTrackBorder = ''},
          },
        },
      },
    }) => unCheckedTrackBorder};
  }
  .MuiSwitch-switchBase.Mui-checked .MuiSwitch-thumb {
    border: ${({
      theme: {
        designSystem: {
          organisms: {
            formSwitch: {checkedTrackBorder = ''},
          },
        },
      },
    }) => checkedTrackBorder};
  };
`;

export default SwitchField;
