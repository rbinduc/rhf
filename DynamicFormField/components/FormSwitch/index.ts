export {default} from './FormSwitch.component';