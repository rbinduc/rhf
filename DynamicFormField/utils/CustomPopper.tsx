import React from 'react';
import {Popper, PopperProps} from '@material-ui/core';

export const CustomPopper = ({placement = 'bottom', ...rest}: PopperProps) => <Popper {...rest} placement={placement} />;
