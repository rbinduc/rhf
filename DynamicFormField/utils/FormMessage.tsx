import styled from '@emotion/styled';
import {FormHelperText} from '@material-ui/core';
import React from 'react';
import {IconNames} from '../../../../icons/IconNameConstants';
import {TenantVector} from '../../../../icons/TenantVector/TenantVector.component';

interface IMessageProps {
  message?: React.ReactNode;
  isBold?: boolean;
  isLarge?: boolean;
}

export interface IFormMessageProps extends IMessageProps {
  id?: string;
  Icon?: IconNames;
  component?: React.ElementType;
  margin?: 'dense';
  isError?: boolean;
  hideIcon?: boolean;
}

const FormMessage = ({id = '', Icon, message, component, margin, isBold, isLarge, hideIcon, isError = true}: IFormMessageProps) =>
  message ? (
    <FormHelperText id={id} error={!!isError} component={component || 'div'} margin={margin}>
      <FormMessageWrapper>
        {!hideIcon && (
          <span id={id + '_icon'}>
            <TenantVector tenantVectorName={Icon || (isError ? IconNames.FieldError : IconNames.Info)} />
          </span>
        )}
        <Message id={id + '_message'} isBold={!!isBold} isLarge={!!isLarge}>
          {message}
        </Message>
      </FormMessageWrapper>
    </FormHelperText>
  ) : null;

const FormMessageWrapper = styled.span`
  display: flex;
  flex-direction: row;
  & svg {
    margin-top: 0.2rem;
    margin-right: 0.5rem;
  }
  word-break: break-word;
`;

const Message = styled.span<IMessageProps>`
  font-weight: ${({isBold}) => (isBold ? 'bold' : 400)};
  font-size: ${({isLarge}) => (isLarge ? '1rem' : '0.875rem')};
`;

export default FormMessage;
