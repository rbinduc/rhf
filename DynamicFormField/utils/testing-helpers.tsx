import React from 'react';
import {render} from '@testing-library/react';
import {FormProvider} from 'react-hook-form';
import useFormHook from './hooks/useFormHook';
import {ThemeProvider} from 'emotion-theming';
import {themeMock} from '../../../../themes/theme-mock';

interface IDefaultValue {
  [key: string]: any;
}

interface RenderOptions {
  defaultValues?: IDefaultValue;
  toPassBack?: any[];
}

/**
 * Testing Library utility function to wrap tested component in React Hook Form
 * @param {ReactElement} ui A React component
 * @param objectParameters
 * @param {Object} objectParameters.defaultValues Initial form values to pass into
 * React Hook Form, which you can then assert against
 * @param {Array} objectParameters.toPassBack React Hook Form method names which we'd
 * like to pass back and use in tests.  A primary use case is sending back 'setError',
 * so we can manually setErrors on React Hook Form components and test error handling
 */
export function renderWithDyanmicForm(
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  {defaultValues = {}, toPassBack = []}: RenderOptions = {}
) {
  let dynamicFormMethods = {} as any;

  const Wrapper = ({children}: {children: React.ReactNode}) => {
    const {methods}: any = useFormHook({defaultValues});
    for (let dynamicFormItem of toPassBack) {
      dynamicFormMethods[dynamicFormItem] = methods[dynamicFormItem];
    }

    return (
      <ThemeProvider theme={themeMock}>
        <FormProvider {...methods}>{children}</FormProvider>
      </ThemeProvider>
    );
  };

  return {
    ...render(ui, {wrapper: Wrapper as React.ComponentType<{}>}),
    dynamicFormMethods,
  };
}

/**
 * Higher order helper function which wraps a component w/ React Hook Form
 * @param {React Component} WrappedComponent to pass into
 * @param {*} restProps any other remaining props
 * @returns {React Component}
 */
export function withDynamicForm(WrappedComponent: React.ComponentType<any>, restProps: object, defaultValues: IDefaultValue) {
  return () => {
    const methods = useFormHook({defaultValues});

    return (
      <ThemeProvider theme={themeMock}>
        <WrappedComponent {...restProps} {...methods} />
      </ThemeProvider>
    );
  };
}
