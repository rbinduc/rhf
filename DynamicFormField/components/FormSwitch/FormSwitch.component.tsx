import * as React from 'react';
import styled from '@emotion/styled';
import {FormControl, InputLabel} from '@material-ui/core';
import {Controller, useFormContext} from 'react-hook-form';
import {ITenantTheme} from '../../../../../../models';
import {IDynamicFormFieldMode, IValidations} from '../../DynamicFormField';
import FormMessage from '../../utils/FormMessage';
import ReactHtmlParser from 'react-html-parser';
import SwitchField, {ISwitchFieldProps} from './SwitchField';

export interface IFormSwitchProps extends ISwitchFieldProps {
  rootWidth?: string;
  rootHeight?: string;
  rootPadding?: string;
  onChange?: (e: any) => void | Promise<void>;
  switchOnLabel?: string;
  switchOffLabel?: string;
  theme?: ITenantTheme;
  description?: string;
  validations?: IValidations;
  isRequired?: boolean;
  isDisabled?: boolean;
  labelId?: string;
  mode?: IDynamicFormFieldMode;
}

/**
 * TODO: We need to refactor structure once more and then we can move this components from organisms to molecules.
 * @param param0
 * @returns
 */
const FormSwitch: React.FC<IFormSwitchProps> = ({
  name,
  size = 'medium',
  onChange,
  label,
  mode,
  description,
  validations,
  defaultValue,
  isRequired,
  isDisabled,
  labelId,
  ...other
}) => {
  const {
    control,
    formState: {errors},
  } = useFormContext();

  /**
   * We need custom logic to check if switch is required since isRequired is not enough,
   * so we need to check if value is actually set to boolean value 'true'.
   */
  const rules = {
    ...validations,
    validate: (value: boolean) => (isRequired ? value : true),
  };

  const errorMessage = errors[name!]?.message;
  return (
    <ContainerStyled>
      <BodyStyled>
        <InputLabelStyled htmlFor={name} id={labelId} fontWeight={description ? 700 : 400}>
          {label}
        </InputLabelStyled>
        <FormControl error={Boolean(errorMessage)} required={isRequired}>
          <Controller
            render={({field}) => (
              <SwitchField checked={field.value} label={label} field={field} isDisabled={isDisabled} size={size} {...other} />
            )}
            name={name as any}
            control={control}
            rules={rules}
            defaultValue={defaultValue}
          />
        </FormControl>
        <FormMessage message={errorMessage} margin="dense" />
      </BodyStyled>
      <FooterStyled>
        <FormSwitchTextStyled shrink={mode === IDynamicFormFieldMode.DENSE}>{ReactHtmlParser(description ?? '')}</FormSwitchTextStyled>
      </FooterStyled>
    </ContainerStyled>
  );
};

interface IFormSwitchLabelStyled {
  shrink?: boolean;
}

const FormSwitchTextStyled = styled.span<IFormSwitchLabelStyled>`
  font-size: ${(props) => (props.shrink ? '0.875rem' : '1rem')};
`;

const ContainerStyled = styled.div`
  display: flex;
  flex-direction: column;
  gap: 1rem;
`;

const BodyStyled = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const FooterStyled = styled.div``;

interface IInputLabelProps {
  fontWeight: number;
}

const InputLabelStyled = styled(InputLabel)<IInputLabelProps>`
  & {
    font-size: 1rem;
    font-weight: ${({fontWeight}) => fontWeight};
  }
`;

export default FormSwitch;
