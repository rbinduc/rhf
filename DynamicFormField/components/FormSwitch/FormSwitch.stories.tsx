import React from 'react';
import {storiesOf} from '@storybook/react';
import FormSwitch from '.';
import {IFormSwitchProps} from './FormSwitch.component';
import {ThemeProvider} from 'emotion-theming';
import {themeMock} from '../../../../../themes/theme-mock';
import {FormProvider} from 'react-hook-form';
import useFormHook from '../../utils/hooks/useFormHook';

storiesOf('Organisms/RHF/DynamicForm/FormSwitch', module)
  .addDecorator((getStory) => <ThemeProvider theme={themeMock}>{getStory()}</ThemeProvider>)
  .add('Default', () => {
    const defaultProps: IFormSwitchProps = {
      id: 'switch_id',
      name: 'tou',
      isRequired: true,
      label: 'Terms of use and Privacy notice*',
      description: 'I acknowledge that I have read and accepted the Terms and Conditions and Privacy Policy',
    };
    const {methods}: any = useFormHook({});
    return (
      <FormProvider {...methods}>
        <FormSwitch {...defaultProps} />
      </FormProvider>
    );
  });
