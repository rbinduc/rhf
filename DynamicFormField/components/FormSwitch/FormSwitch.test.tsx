import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import FormSwitch from '.';
import {IFormSwitchProps} from './FormSwitch.component';
import {renderWithDyanmicForm} from '../../utils/testing-helpers';

const defaultProps: IFormSwitchProps = {
  id: 'switch_id',
  name: 'tou',
  isRequired: true,
  label: 'Terms of use and Privacy notice*',
  description: 'I acknowledge that I have read and accepted the Terms and Conditions and Privacy Policy',
};

jest.mock('@ionic-native/keyboard', () => ({
  hide: jest.fn(),
}));

jest.mock('../../../../../../automation/PageLocators', () => ({
  locators: {
    jobPageLocators: {
      aFiltersIcon: 'filtersIcon',
    },
    commonLocators: {
      aHeaderText: 'headerText',
    },
    accountSettings: {
      aChangeCountryHeader: 'changeCountryHeader',
    },
  },
}));

describe('FormSwitch', () => {
  it('snapshot match', () => {
    const {container} = renderWithDyanmicForm(<FormSwitch {...defaultProps} />, {});
    expect(container).toBeDefined();
    expect(container.firstChild).toMatchSnapshot();
  });
});
