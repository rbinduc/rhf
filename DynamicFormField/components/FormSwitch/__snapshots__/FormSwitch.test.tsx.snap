// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSwitch snapshot match 1`] = `
<div
  class="css-1uxw6he-ContainerStyled e17m884d1"
>
  <div
    class="css-8jys7u-BodyStyled e17m884d2"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root css-1q2zhaq-InputLabelStyled e17m884d4 MuiInputLabel-animated"
      font-weight="700"
      for="tou"
    >
      Terms of use and Privacy notice*
    </label>
    <div
      class="MuiFormControl-root"
    >
      <span
        class="MuiSwitch-root css-1kcy543-StyledSwitch e1mmj7rs0"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiSwitch-switchBase MuiSwitch-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              aria-label="Terms of use and Privacy notice*"
              class="PrivateSwitchBase-input-4 MuiSwitch-input"
              id="switch_id"
              type="checkbox"
              value=""
            />
            <span
              class="MuiSwitch-thumb"
            />
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiSwitch-track"
        />
      </span>
    </div>
  </div>
  <div
    class="css-t8o0nl-FooterStyled e17m884d3"
  >
    <span
      class="css-1ji3mfx-FormSwitchTextStyled e17m884d0"
    >
      I acknowledge that I have read and accepted the Terms and Conditions and Privacy Policy
    </span>
  </div>
</div>
`;
